payment-guarantee/src/contract.rs

```diff
@@ -0,0 +1,196 @@
+use crate::query::QueryMsg;
+use crate::state::CONTRACT_INFO;
+use crate::types::ContractInfo;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let contract_info = ContractInfo {
+        source_contracts: msg.source_contracts,
+        payer: info.sender.to_string(),
+    };
+
+    CONTRACT_INFO.save(deps.storage, &contract_info);
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Deposit {} => execute_deposit(deps, env, info),
+        ExecuteMsg::ReceiveNft(msg) => execute_receive_nft(deps, env, info, msg),
+    }
+}
+
+pub fn execute_deposit(
+    mut _deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    if info.funds.is_empty() {
+        return Err(ContractError::InvalidParameter { msg: "amount is empty.".to_string() });
+    }
+
+    Ok(Response {
+        submessages: vec![],
+        messages: vec![],
+        attributes: vec![
+            attr("action", "deposit"),
+            attr("sender", info.sender),
+        ],
+        data: None,
+    })
+}
+
+pub fn execute_receive_nft(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let source_contract = info.sender.clone();
+    let contract_info = CONTRACT_INFO.load(deps.storage)?;
+    if is_invalid_from_contract(&contract_info, source_contract.to_string()) {
+        return Err(ContractError::Unauthorized {
+            msg: format!("The token has a different contract. source: {}, this: {}", source_contract.as_str(), contract_info.source_contracts.join(",")),
+        });
+    }
+
+    let token_id = msg.token_id.to_string();
+    let query_msg = QueryMsg::OwnerOf {
+        token_id: token_id.to_string(),
+        include_expired: None,
+    };
+
+    let res: OwnerOfResponse = deps.querier.query_wasm_smart(source_contract.as_str(), &query_msg)?;
+    if env.contract.address.to_string().ne(res.owner.as_str()) {
+        return Err(ContractError::Unauthorized { msg: "The owner of the token must be this contract.".to_string() });
+    }
+
+    let res: StdResult<TokenInfoMsg> = from_binary(&msg.msg.unwrap());
+    if res.is_err() {
+        return Err(ContractError::InvalidParameter { msg: "msg is invalid parameter.".to_string() });
+    }
+
+    let token_info_msg = res.unwrap();
+    let token_price = token_info_msg.price;
+
+    let deposit_coin = deps.querier.query_balance(env.contract.address, token_price.denom.clone())?;
+    if deposit_coin.amount.le(token_price.amount.borrow()) {
+        return Err(ContractError::InsufficientDeposit {});
+    }
```

Comment:
If the current Contract does not have enough deposit, the transaction will fail.
The ability to change the owner of NFT to an existing owner should be added later.